### PR TITLE
Backport fixes to 13.0 for debian testing

### PR DESCRIPTION
**Pull request by d-fence**
The current "testing" (and future stable) updates various packages to versions which are not fully backwards-compatible and break Odoo 13, notably:

* [Python 3.8.4](https://packages.debian.org/bullseye/python3.8) [raises a `ValueError` when `ast.Name` is used with the constants `True`, `False` or `None`](https://docs.python.org/3/whatsnew/changelog.html#python-3-8-4-release-candidate-1) (possibly this is silently broken in previous Python 3.8 versions)
* [Werkzeug 1.0](https://packages.debian.org/bullseye/python3-werkzeug) breaks all previous deprecation warnings:
  - removes various aliases
  - moves packages around
  - removes some packages entirely

### CODE REVIEW

**Comment by xmo-odoo** [comment]
That's a lot.

Also looking at the ast doc, `NameConstant` doesn't have a ctx.

**Comment by d-fence** [comment]
> That's a lot.
> 
> Also looking at the ast doc, `NameConstant` doesn't have a ctx.

Hmm, I'm wondering if we should replace it directly by Constant instead as NameConstant is [already deprecated](https://greentreesnakes.readthedocs.io/en/latest/nodes.html#NameConstant) and Constant [exists since 3.6](https://greentreesnakes.readthedocs.io/en/latest/nodes.html#Constant)

**Comment by xmo-odoo** [comment]
If it works correctly in 3.6 and it's only done for 13.0 then fine with me.

**Comment by d-fence** [comment]
> If it works correctly in 3.6 and it's only done for 13.0 then fine with me.

This sentence about Constant is not clear:
`This class is available in the ast module from Python 3.6, but it isn’t produced by parsing code until Python 3.8.`

Not sure that it will work in 3.6. I'll test.

**Comment by xmo-odoo** [comment]
> > If it works correctly in 3.6 and it's only done for 13.0 then fine with me.
> 
> This sentence about Constant is not clear:
> `This class is available in the ast module from Python 3.6, but it isn’t produced by parsing code until Python 3.8.`

I think it just means that before 3.8 `ast.parse` would always generate the "old" constant encoding.

File: odoo/addons/base/models/ir_qweb.py [hunk 1]
```diff
@@ -426,10 +426,10 @@ def _compile_expr(self, expr):
...
+                return ast.Constant(True)
             attr = attr.lower()
             if attr in ('false', '0'):
-                return ast.Name(id='False', ctx=ast.Load())
+                return ast.Constant(False)
             elif attr in ('true', '1'):
-                return ast.Name(id='True', ctx=ast.Load())
-        return ast.Name(id=str(attr if attr is False else default), ctx=ast.Load())
+                return ast.Constant(True)
+        return ast.Constant(attr if attr is False else default)
```
[the remark this answers is not in the document]
Comment (d-fence) [reply odoo/addons/base/models/ir_qweb.py:435]: No, Constant does not need a ctx: `Constant(constant value, string? kind)`

I agree that the line is weird and in practice the only values should be False or True but it shorten the if/elif.

[the remark this answers is not in the document]
Comment (odony) [reply odoo/addons/base/models/ir_qweb.py:435]: > No, Constant does not need a ctx: `Constant(constant value, string? kind)`

Obviously it doesn't, it's a constant :p The question is that we're replacing a `Name` that did have access to a ctx with a `Constant`, and we're not sure what `default` can hold ;-)

I was hoping you had any idea (e.g. through testing) what cases were possible for `default` or if it was used at all. This is the only part of the patch that seems risky for stable (remember, we're fixing a problem that doesn't exist for the majority of our users, so it *must not* break anything in existing deployments).
If we're unsure, we could do the change in master and wait for it to be tested by devs for a few weeks, and *then* backport it to 13.0.

[the remark this answers is not in the document]
Comment (odony) [reply odoo/addons/base/models/ir_qweb.py:435]: Perhaps better with an example ;-)
```py
In [5]: this = "that"                                                                                                                                                                                                       

In [6]: e = ast.fix_missing_locations(ast.Expression(ast.Name(id='this', ctx=ast.Load())))                                                                                                                                  

In [7]: eval(compile(e, "<>", "eval"))                                                                                                                                                                                      
Out[7]: 'that'

In [8]: e = ast.fix_missing_locations(ast.Expression(ast.Constant('this')))                                                                                                                                                 

In [9]: eval(compile(e, "<>", "eval"))                                                                                                                                                                                      
Out[9]: 'this'
```

[the remark this answers is not in the document]
Comment (xmo-odoo) [reply odoo/addons/base/models/ir_qweb.py:435]: > > No, Constant does not need a ctx: `Constant(constant value, string? kind)`
> 
> Obviously it doesn't, it's a constant :p The question is that we're replacing a `Name` that did have access to a ctx with a `Constant`, and we're not sure what `default` can hold ;-)
> 
> I was hoping you had any idea (e.g. through testing) what cases were possible for `default` or if it was used at all. This is the only part of the patch that seems risky for stable (remember, we're fixing a problem that doesn't exist for the majority of our users, so it _must not_ break anything in existing deployments).

There are [only 5 callsites for `_get_attr_bool`](https://github.com/odoo/odoo/blob/79a7cf9be53197d82167bec19bf971c9aa0804dc/odoo/addons/base/models/ir_qweb.py#L205-L218) and none of them uses an explicit `default`. Either way for safety we could just `bool(default)` or `assert default in (True, False)`. The point of the method is to return a boolean from whatever the input is, so passing in a non-bool default doesn't really make sense.

**Comment by elmeriniemela** [comment]
I request that this gets back ported to 11.0 and 12.0 as well so it would be best if the fix worked on older versions of python as well. The commit on #55027 works with every python version.

**Comment by xmo-odoo** [comment]
@d-fence @odony cherry-picked the various werkzeug fixes from master as-is and in-order.

File: addons/website/models/res_partner.py [hunk 2]
```diff
@@ -27,14 +22,14 @@ def google_map_img(self, zoom=8, width=298, height=298):
...
-        return urlplus('//maps.googleapis.com/maps/api/staticmap', params)
+        return '//maps.googleapis.com/maps/api/staticmap?'+werkzeug.urls.url_encode(params)
 
     def google_map_link(self, zoom=10):
         params = {
             'q': '%s, %s %s, %s' % (self.street or '', self.city or '', self.zip or '', self.country_id and self.country_id.display_name or ''),
             'z': zoom,
         }
-        return urlplus('https://maps.google.com/maps', params)
+        return 'https://maps.google.com/maps' + werkzeug.urls.url_encode(params)
```
Comment (xmo-odoo) [inline addons/website/models/res_partner.py:32]: FIXME! forgot the `?`

**Comment by 0fisher** [comment]
python3.8.5 raise 
```Shell
Traceback (most recent call last):
  File "/Users/0fisher/workspace/odoo13/odoo/addons/http_routing/models/ir_http.py", line 658, in _handle_exception
    html = cls._get_error_html(env, code, values)
  File "/Users/0fisher/workspace/odoo13/odoo/addons/http_routing/models/ir_http.py", line 603, in _get_error_html
    return env['ir.ui.view'].render_template('http_routing.%s' % code, values)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/ir_ui_view.py", line 1177, in render_template
    return self.browse(self.get_view_id(template)).render(values, engine)
  File "/Users/0fisher/workspace/odoo13/odoo/addons/web_editor/models/ir_ui_view.py", line 27, in render
    return super(IrUiView, self).render(values=values, engine=engine, minimal_qcontext=minimal_qcontext)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/ir_ui_view.py", line 1185, in render
    return self.env[engine].render(self.id, qcontext)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/ir_qweb.py", line 58, in render
    result = super(IrQWeb, self).render(id_or_xml_id, values=values, **context)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/qweb.py", line 260, in render
    self.compile(template, options)(self, body.append, values or {})
  File "<decorator-gen-53>", line 2, in compile
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/tools/cache.py", line 90, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/ir_qweb.py", line 113, in compile
    return super(IrQWeb, self).compile(id_or_xml_id, options=options)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/qweb.py", line 323, in compile
    raise QWebException("Error when compiling AST", e, path, node and etree.tostring(node[0], encoding='unicode'), name)
odoo.addons.base.models.qweb.QWebException: Name node can't be used with 'None' constant
Traceback (most recent call last):
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/tools/cache.py", line 85, in lookup
    r = d[key]
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/tools/func.py", line 69, in wrapper
    return func(self, *args, **kwargs)
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/tools/lru.py", line 44, in __getitem__
    a = self.d[obj].me
KeyError: ('ir.qweb', <function IrQWeb.compile at 0x110a52e50>, 394, ('zh_CN', None, None, None, False, None))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/0fisher/workspace/odoo13/odoo/odoo/addons/base/models/qweb.py", line 315, in compile
    unsafe_eval(compile(astmod, '<template>', 'exec'), ns)
ValueError: Name node can't be used with 'None' constant

Error when compiling AST
ValueError: Name node can't be used with 'None' constant
Template: http_routing.404
Path: /t/t/div/div/div/ul/li/a
Node: <a href="/my">文档</a>
```

**Comment by jeroen7s** [comment]
Ubuntu20.04 just started upgrading its built-in python from version 3.8.2 to 3.8.5, making this AST fix more urgent for many people running this OS on the server.

**Comment by odony** [comment]
> Ubuntu20.04 just started upgrading its built-in python from version 3.8.2 to 3.8.5, making this AST fix more urgent for many people running this OS on the server.

May I ask where you got that info from? https://packages.ubuntu.com/focal/python3 still has 3.8.2, just like `apt-cache policy` :thinking:

**Comment by Yenthe666** [comment]
@odony I'm not sure either but my latest apt-get update on my local system just broke Odoo completely. This Linux is now Python 3.8.5 on Ubuntu 20.04.1 LTS.

**Comment by jeroen7s** [comment]
Not sure why that URL still shows python version 3.8.2, but me and some of my colleagues are definitely getting 3.8.5 from the Ubuntu repos, I actually got it yesterday already.

**Comment by robin5210** [comment]
It doesn't come from the main focal repo, but instead from the upgrade or security repo.

python3.8/focal-updates,focal-security 3.8.5-1~20.04 amd64 [upgradable from: 3.8.2-1ubuntu1.2]

**Comment by Yenthe666** [comment]
Good catch @robin5210! This also means that anyone running Ubuntu 20.04.x doing an update will have completely broken db's.

**Comment by odony** [comment]
Confirmed, seems like the python3.8 package itself got upgraded:  https://packages.ubuntu.com/focal/python3.8
It's getting installed via unattended upgrades indeed, so if you restart your odoo service afterwards you're in trouble.

# grep -C2 "libpython.*3.8.5" /var/log/apt/history.log
Start-Date: 2020-10-08  06:31:38
Commandline: /usr/bin/unattended-upgrade
Upgrade: libpython3.8-dev:amd64 (3.8.2-1ubuntu1.2, 3.8.5-1~20.04), libpython3.8-minimal:amd64 (3.8.2-1ubuntu1.2, 3.8.5-1~20.04), libpython3.8:amd64 (3.8.2-1ubuntu1.2, 3.8.5-1~20.04), python3.8:amd64 (3.8.2-1ubuntu1.2, 3
.8.5-1~20.04), python3.8-minimal:amd64 (3.8.2-1ubuntu1.2, 3.8.5-1~20.04), libpython3.8-stdlib:amd64 (3.8.2-1ubuntu1.2, 3.8.5-1~20.04), python3.8-dev:amd64 (3.8.2-1ubuntu1.2, 3.8.5-1~20.04)
End-Date: 2020-10-08  06:31:41

**Comment by pedrobaeza** [comment]
@odony (and all), please read my comment at https://github.com/odoo/odoo/issues/59456#issuecomment-705433473

**Comment by odony** [comment]
> @odony (and all), please read my comment at [#59456 (comment)](https://github.com/odoo/odoo/issues/59456#issuecomment-705433473)

I mostly agree with you :) But I also think our policy must be driven by common sense, just like we do for other bugfixes and backports.

For sure our main goal remains compatibility with the versions in the Debian/Ubuntu LTS _at the time of the release_. So for v13 that's clearly bionic/buster, which means 3.7 and 3.6 only.
But if we notice that there is a major incompatibility with the default Python version in a newer LTS versions, and the fix is not too risky, I think it is worth a backport, because it will save many users a bad experience and lots of wasted time looking up error messages online. And of course it will save us duplicate issues and tickets ;-)

It is always a question of balance between risk and benefit. For instance this PR contains too many risky changes, so we're proceeding with only backporting the necessary AST fix via #59480. See also the rationale on that PR, and the https://github.com/odoo/odoo/pull/59480#issuecomment-705557805

PS: bugs caused by changes in `requirements.txt` are indeed annoying but less problematic than real incompatibilities with a core library update. In most cases they will only affect new installations, whereas unattended security upgrades of Python distribution on Ubuntu LTS could cause production downtime without any human intervention.

**Comment by mk298** [comment]
I am having the same issue while running odoo 13 in my local system. Can anyone suggest how to fix it? should I downgrade the python version ?

**Comment by Yenthe666** [comment]
@mk298 downgrade or apply the patch from https://github.com/odoo/odoo/pull/59480 locally while Odoo works on the PR to get merged in.

**Comment by mk298** [comment]
> @mk298 downgrade or apply the patch from #59480 locally while Odoo works on the PR to get merged in.

@Yenthe666 Thank you

**Comment by TheCloneMaster** [comment]
Patch https://github.com/odoo/odoo/pull/59480 tested and working !!

**Comment by omalbastin** [comment]
I am having the same issue in v12.

**Comment by d-fence** [comment]
partly done in #59480.

**Pull request closed without merging**